File: SOC_Photon/src/Coulombs.cpp
//
// MIT License
//
// Copyright (C) 2023 - Dave Gutz
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include "application.h"
#include <math.h>
#include "Battery.h"
#include "Coulombs.h"
#include "parameters.h"
extern SavedPars sp;      // Various parameters to be static at system level and saved through power cycle
#include "command.h"
extern CommandPars cp;    // Various parameters to be static at system level
extern PublishPars pp;    // For publishing


// class Coulombs
Coulombs::Coulombs() {}
Coulombs::Coulombs(double *sp_delta_q, float *sp_t_last, const float q_cap_rated, const float t_rlim,
  uint8_t *sp_mod_code, const double s_coul_eff)
  : q_(q_cap_rated), q_capacity_(q_cap_rated), q_cap_rated_(q_cap_rated), q_cap_rated_scaled_(q_cap_rated),
    q_min_(0.), sat_(true), soc_(1.), soc_min_(0.), sp_delta_q_(sp_delta_q), sp_t_last_(sp_t_last), t_rlim_(0.017),
    chem_(sp_mod_code)
    {
      coul_eff_ = (chem_.coul_eff*s_coul_eff);
    }
Coulombs::~Coulombs() {}


// operators
// Pretty print
void Coulombs::pretty_print(void)
{
  Serial.printf("Coulombs:\n");
  Serial.printf(" coul_eff%9.5f\n", coul_eff_);
  Serial.printf(" delta_q%9.1f, C\n", *sp_delta_q_);
  Serial.printf(" mod %s\n", chem_.decode(mod_code()).c_str());
  Serial.printf(" mod_code %d\n", mod_code());
  Serial.printf(" q%9.1f, C\n", q_);
  Serial.printf(" q_cap%9.1f, C\n", q_capacity_);
  Serial.printf(" q_cap_rat%9.1f, C\n", q_cap_rated_);
  Serial.printf(" q_cap_rat_scl%9.1f, C\n", q_cap_rated_scaled_);
  Serial.printf(" q_min%9.1f, C\n", q_min_);
  Serial.printf(" resetting %d\n", resetting_);
  Serial.printf(" sat %d\n", sat_);
  Serial.printf(" soc%8.4f\n", soc_);
  Serial.printf(" soc_min%8.4f\n", soc_min_);
  Serial.printf(" t_last%5.1f dg C\n", *sp_t_last_);
  Serial.printf(" rated_t%5.1f dg C\n", chem_.rated_temp);
  Serial.printf(" t_rlim%7.3f dg C / s\n", t_rlim_);
  Serial.printf("Coulombs::");
  chem_.pretty_print();
}

// functions

// Scale size of battery and adjust as needed to preserve delta_q.  Tb unchanged.
// Goal is to scale battery and see no change in delta_q on screen of 
// test comparisons.   The rationale for this is that the battery is frequently saturated which
// resets all the model parameters.   This happens daily.   Then both the model and the battery
// are discharged by the same current so the delta_q will be the same.
void Coulombs::apply_cap_scale(const float scale)
{
  q_cap_rated_scaled_ = scale * q_cap_rated_;
  q_capacity_ = calculate_capacity(*sp_t_last_);
  q_ = *sp_delta_q_ + q_capacity_; // preserve delta_q, deficit since last saturation (like real life)
  soc_ = q_ / q_capacity_;
  resetting_ = true;     // momentarily turn off saturation check
}

// Memory set, adjust book-keeping as needed.  delta_q, capacity, temp preserved
void Coulombs::apply_delta_q(const double delta_q)
{
  *sp_delta_q_ = delta_q;
  q_ = *sp_delta_q_ + q_capacity_;
  soc_ = q_ / q_capacity_;
  resetting_ = true;     // momentarily turn off saturation check
}

// Memory set, adjust book-keeping as needed.  q_cap_ etc presesrved
void Coulombs::apply_delta_q_t(const boolean reset)
{
  if ( !reset ) return;
  q_capacity_ = calculate_capacity(*sp_t_last_);
  q_ = q_capacity_ + *sp_delta_q_;
  soc_ = q_ / q_capacity_;
  resetting_ = true;
}
void Coulombs::apply_delta_q_t(const double delta_q, const float temp_c)
{
  *sp_delta_q_ = delta_q;
  *sp_t_last_ = temp_c;
  apply_delta_q_t(true);
}


// Memory set, adjust book-keeping as needed.  delta_q preserved
void Coulombs::apply_soc(const float soc, const float temp_c)
{
  soc_ = soc;
  q_capacity_ = calculate_capacity(temp_c);
  q_ = soc*q_capacity_;
  *sp_delta_q_ = q_ - q_capacity_;
  resetting_ = true;     // momentarily turn off saturation check
}

// Capacity
double Coulombs::calculate_capacity(const float temp_c)
{
  return( q_cap_rated_scaled_ * (1+chem_.dqdt*(temp_c - chem_.rated_temp)) );
}

/* Coulombs::count_coulombs:  Count coulombs based on true=actual capacity
Inputs:
  dt              Integration step, s
  temp_c          Battery temperature, deg C
  charge_curr     Charge, A
  sat             Indication that battery is saturated, T=saturated
  tlast           Past value of battery temperature used for rate limit memory, deg C
  coul_eff        Coulombic efficiency - the fraction of charging input that gets turned into usable Coulombs
Outputs:
  q_capacity_     Saturation charge at temperature, C
  *sp_delta_q_    Charge change since saturated, C
  *sp_t_last_     Updated value of battery temperature used for rate limit memory, deg C
  resetting_      Sticky flag for initialization, T=reset
  soc_            Fraction of saturation charge (q_capacity_) available (0-1) 
  soc_min_        Estimated soc where battery BMS will shutoff current, fraction
  q_min_          Estimated charge at low voltage shutdown, C\
*/
float Coulombs::count_coulombs(const double dt, const boolean reset_temp, const float temp_c, const float charge_curr,
  const boolean sat, const double delta_q_ekf)
{
    // Rate limit temperature.   When modeling, reset_temp.  In real world, rate limited Tb ramps Coulomb count since bms_off
    if ( reset_temp && sp.mod_vb() )
    {
      *sp_t_last_ = temp_c;
    }
    float temp_lim = max(min( temp_c, *sp_t_last_ + t_rlim_*dt), *sp_t_last_ - t_rlim_*dt);

    // State change
    double d_delta_q = charge_curr * dt;
    if ( charge_curr>0. ) d_delta_q *= coul_eff_;
    d_delta_q -= chem_.dqdt*q_capacity_*(temp_lim - *sp_t_last_);
    sat_ = sat;

    // Saturation.   Goal is to set q_capacity and hold it so remember last saturation status.
    if ( sat )
    {
        if ( d_delta_q > 0 )
        {
            d_delta_q = 0.;
            if ( !resetting_ )
            {
              *sp_delta_q_ = 0.;
            }
        }
        else if ( reset_temp )
        {
          *sp_delta_q_ = 0.;
        }
    }
    // else if ( reset_temp && !cp.fake_faults ) *sp_delta_q_ = delta_q_ekf;  // Solution to booting up unsaturated
    resetting_ = false;     // one pass flag

    // Integration.   Can go to negative
    q_capacity_ = calculate_capacity(temp_lim);
    if ( !reset_temp )
    {
      *sp_delta_q_ = max(min(*sp_delta_q_ + d_delta_q, 0.0), -q_capacity_*1.5);
    }
    q_ = q_capacity_ + *sp_delta_q_;

    // Normalize
    soc_ = q_ / q_capacity_;
    soc_min_ = chem_.soc_min_T_->interp(temp_lim);
    q_min_ = soc_min_ * q_capacity_;

    // if ( sp.debug==96 )
    //     Serial.printf("cc:,reset_temp,dt,voc, Voc_filt, V_sat, temp_c, temp_lim, sat, charge_curr, d_d_q, d_q, q, q_cap, soc, %d,%7.3f,%7.3f,%7.3f,%7.3f,%7.3f,%7.3f,%d,%7.3f,%10.6f,%9.1f,%9.1f,%9.1f,%7.4f,\n",
    //                 reset, dt, pp.pubList.Voc, pp.pubList.Voc_filt,  this->Vsat(), temp_c, temp_lim, sat, charge_curr, d_delta_q, *sp_delta_q_, q_, q_capacity_, soc_);

    // Save and return
    *sp_t_last_ = temp_lim;
    return ( soc_ );
}

File: SOC_Photon/py/Coulombs.py
# BatteryEKF - general purpose battery class for embedded EKF
# Copyright (C) 2022 Dave Gutz
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation;
# version 2.1 of the License.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# See http://www.fsf.org/licensing/licenses/lgpl.txt for full license text.

"""Coulomb Counting.  1 Ampere is 1 Coulombs/second charge rate of change.   Herein are various methods to keep track
of the totals and standardize the calculations."""

# Constants
import numpy as np
from pyDAGx import myTables
from Chemistry_BMS import Chemistry, BMS

dqdt = 0.01  # Change of charge with temperature, fraction/deg C.  From literature.  0.01 is commonly used
coul_eff = 0.9985  # Coulombic efficiency - the fraction of charging input that gets turned into usable Coulombs


class Coulombs:
    """Coulomb Counting"""

    def __init__(self, q_cap_rated, q_cap_rated_scaled, t_rated, t_rlim=0.017, tweak_test=False,
                 coul_eff_= 0.9985, mod_code=0):
        self.q_cap_rated = q_cap_rated
        self.q_cap_rated_scaled = q_cap_rated_scaled
        self.t_rated = t_rated
        self.t_rlim = t_rlim
        self.delta_q = 0.
        self.q = 0.
        self.q_capacity = 0.
        self.soc_min = 0.
        self.soc = 1.
        self.resetting = True
        self.q_min = 0.
        self.temp_lim = 0.
        self.t_last = 0.
        self.sat = True
        self.chm = mod_code
        self.coul_eff = coul_eff_
        self.tweak_test = tweak_test
        self.reset = False
        self.chemistry = Chemistry(mod_code=mod_code)

    def __str__(self, prefix=''):
        """Returns representation of the object"""
        s = prefix + "Coulombs:\n"
        s += "  q_cap_rated = {:9.1f}    // Rated capacity at t_rated_, saved for future scaling, C\n"\
            .format(self.q_cap_rated)
        s += "  q_cap_rated_scaled = {:9.1f} // Applied rated capacity at t_rated_, after scaling, C\n"\
            .format(self.q_cap_rated_scaled)
        s += "  q_capacity = {:9.1f}     // Saturation charge at temperature, C\n".format(self. q_capacity)
        s += "  q =          {:9.1f}     // Present charge available to use, C\n".format(self. q)
        s += "  delta_q      {:9.1f}     // Charge since saturated, C\n".format(self. delta_q)
        s += "  soc =        {:7.3f}       // Fraction of saturation charge (q_capacity_) available (0-1)  soc)\n"\
            .format(self.soc)
        s += "  sat =          {:1.0f}          // Indication from caller that battery is saturated, T=saturated\n"\
            .format(self.sat)
        s += "  t_rated =    {:5.1f}         // Rated temperature, deg C\n".format(self. t_rated)
        s += "  t_last =     {:5.1f}         // Last battery temperature for rate limit memory, deg C\n"\
            .format(self.t_last)
        s += "  t_rlim =     {:7.3f}       // Tbatt rate limit, deg C / s\n".format(self. t_rlim)
        s += "  resetting =     {:d}          // Flag to test coulomb counters, T = external reset of counter\n"\
            .format(self.resetting)
        s += "  soc_min =    {:7.3f}       // Lowest soc for power delivery.   Arises with temp < 20 C\n"\
            .format(self.soc_min)
        s += "  tweak_test =    {:d}          // Driving signal injection completely using software inj_soft_bias\n"\
            .format(self.tweak_test)
        s += "  coul_eff =   {:8.4f}      // Coulombic efficiency- fraction of charging turned into usable Coulombs\n".\
            format(self.coul_eff)
        return s

    def apply_cap_scale(self, scale):
        """ Scale size of battery and adjust as needed to preserve delta_q.  Tb unchanged.
        Goal is to scale battery and see no change in delta_q on screen of
        test comparisons.   The rationale for this is that the battery is frequently saturated which
        resets all the model parameters.   This happens daily.   Then both the model and the battery
        are discharged by the same current so the delta_q will be the same."""
        self.q_cap_rated_scaled = scale * self.q_cap_rated
        self.q_capacity = self.calculate_capacity(self.t_last)
        self.q = self.delta_q + self.q_capacity  # preserve self.delta_q, deficit since last saturation(like real life)
        self.soc = self.q / self.q_capacity
        self.resetting = True  # momentarily turn off saturation check

    def apply_delta_q(self, delta_q, temp_c):
        """Memory set, adjust bookkeeping as needed.  delta_q, capacity, temp preserved"""
        self.delta_q = delta_q
        self.q_capacity = self.calculate_capacity(temp_c)
        self.q = self.delta_q + self.q_capacity
        self.soc = self.q / self.q_capacity
        self.resetting = True  # momentarily turn off saturation check

    def apply_delta_q_t(self, delta_q, temp_c):
        self.delta_q = delta_q
        self.q_capacity = self.calculate_capacity(temp_c)
        self.q = self.q_capacity + self.delta_q
        self.soc = self.q / self.q_capacity
        self.resetting = True

    def apply_soc(self, soc, temp_c):
        """Memory set, adjust bookkeeping as needed.  delta_q preserved"""
        self.soc = soc
        self.q_capacity = self.calculate_capacity(temp_c)
        self.q = soc*self.q_capacity
        self.delta_q = self.q - self.q_capacity
        self.resetting = True  # momentarily turn off saturation check

    def calculate_capacity(self, temp_c):
        """Capacity"""
        try:
            res = self.q_cap_rated_scaled * (1. + dqdt * (temp_c - self.t_rated))
        except:
            res = 1
        return res
        # return self.q_cap_rated_scaled * (1. + dqdt * (temp_c - self.t_rated))

    def count_coulombs(self, chem, dt, reset, temp_c, charge_curr, sat, soc_init=None):
        """Count coulombs based on true=actual capacity
        Inputs:
            dt              Integration step, s
            temp_c          Battery temperature, deg C
            charge_curr     Charge, A
            sat             Indicator that battery is saturated (VOC>threshold(temp)), T/F
            t_last          Past value of battery temperature used for rate limit memory, deg C
            coul_eff        Coulombic efficiency - the fraction of charging input that gets turned into usable Coulombs
        """
        if self.chm != chem:
            self.chemistry.assign_all_mod(chem)
            self.chm = chem
        self.reset = reset
        d_delta_q = charge_curr * dt
        if charge_curr > 0. and not self.tweak_test:
            d_delta_q *= self.coul_eff
        self.sat = sat

        # Rate limit temperature
        self.temp_lim = max(min(temp_c, self.t_last + self.t_rlim*dt), self.t_last - self.t_rlim*dt)
        if reset:
            self.temp_lim = temp_c
            self.t_last = temp_c

        # Saturation.   Goal is to set q_capacity and hold it so remembers last saturation status.
        if sat:
            if d_delta_q > 0:
                d_delta_q = 0.
                if ~self.resetting:
                    self.delta_q = 0.
        # elif reset:
        #     self.delta_q = 0.
        self.resetting = False  # one pass flag.  Saturation debounce should reset next pass

        # Integration
        self.q_capacity = self.calculate_capacity(self.temp_lim)
        self.delta_q = max(min(self.delta_q + d_delta_q - dqdt*self.q_capacity*(self.temp_lim-self.t_last),
                               0.0), -self.q_capacity*1.5)
        self.q = self.q_capacity + self.delta_q

        # Normalize
        self.soc = self.q / self.q_capacity
        self.soc_min = self.chemistry.lut_min_soc.interp(self.temp_lim)
        self.q_min = self.soc_min * self.q_capacity

        # Save and return
        self.t_last = self.temp_lim
        return self.soc

    def load(self, delta_q, t_last):
        """Load states from retained memory"""
        self.delta_q = delta_q
        self.t_last = t_last
